src/Features/Core/Portable/ChangeSignature/AbstractChangeSignatureService.cs

```diff
@@ -147,35 +171,32 @@ internal ChangeSignatureResult ChangeSignature(Document document, int position,
             }
 
             var solution = document.Project.Solution;
-            var documentId = solution.GetDocumentId(declarationLocation.SourceTree);
-            var declarationDocument = solution.GetDocument(documentId);
-            var declarationChangeSignatureService = declarationDocument?.GetRequiredLanguageService<AbstractChangeSignatureService>();
+            var declarationDocument = solution.GetRequiredDocument(declarationLocation.SourceTree!);
+            var declarationChangeSignatureService = declarationDocument.GetRequiredLanguageService<AbstractChangeSignatureService>();
 
-            if (declarationChangeSignatureService == null)
-            {
-                return new CannotChangeSignatureAnalyzedContext(CannotChangeSignatureReason.DeclarationLanguageServiceNotFound);
-            }
+            int insertPosition;
+            var reference = symbol.DeclaringSyntaxReferences.FirstOrDefault();
 
-            int? insertPositionOpt = declarationChangeSignatureService.TryGetInsertPositionFromDeclaration(symbol.Locations.FirstOrDefault().FindNode(cancellationToken));
-            if (insertPositionOpt == null)
+            if (reference != null)
             {
-                return new CannotChangeSignatureAnalyzedContext(CannotChangeSignatureReason.DeclarationMethodPositionNotFound);
+                insertPosition = declarationChangeSignatureService.GetPositionBeforeParameterListClosingBrace(reference.GetSyntax());
             }
-
-            if (!symbol.MatchesKind(SymbolKind.Method, SymbolKind.Property, SymbolKind.NamedType))
+            else
             {
-                return new CannotChangeSignatureAnalyzedContext(CannotChangeSignatureReason.IncorrectKind);
+                // We may have some indirection, like an event declaration backed by an invisible delegate invoke method.
+                // In the worst case, still continue on.
```

Comment:
not sure what is meant by 'some indirection'.

Reply:
Updated the comment. Not sure if it's any more satisfactory of an answer.